# Description: use libunique so only one instance of the capplet is opened
# Ubuntu: https://bugs.launchpad.net/bugs/531974
# Upstream: 
=== modified file 'gnome-control-center-2.29.91/capplets/about-me/Makefile.am'
Index: gnome-control-center-2.30.0/capplets/about-me/Makefile.am
===================================================================
--- gnome-control-center-2.30.0.orig/capplets/about-me/Makefile.am	2009-08-24 12:54:41.000000000 +0200
+++ gnome-control-center-2.30.0/capplets/about-me/Makefile.am	2010-03-29 17:01:49.013545896 +0200
@@ -31,7 +31,7 @@
 if BUILD_ABOUTME
 bin_PROGRAMS = gnome-about-me
 
-gnome_about_me_LDADD = $(GNOMECC_CAPPLETS_LIBS) $(LIBEBOOK_LIBS)
+gnome_about_me_LDADD = $(GNOMECC_CAPPLETS_LIBS) $(UNIQUE_LIBS) $(LIBEBOOK_LIBS)
 gnome_about_me_LDFLAGS = -export-dynamic
 
 @INTLTOOL_DESKTOP_RULE@
@@ -44,6 +44,7 @@
 
 INCLUDES = \
 	$(GNOMECC_CAPPLETS_CFLAGS) \
+	$(UNIQUE_CFLAGS) \
 	$(LIBEBOOK_CFLAGS) \
 	-DDATADIR="\"$(datadir)\"" \
 	-DGNOMECC_DATA_DIR="\"$(pkgdatadir)\"" \
Index: gnome-control-center-2.30.0/capplets/about-me/gnome-about-me.c
===================================================================
--- gnome-control-center-2.30.0.orig/capplets/about-me/gnome-about-me.c	2010-03-29 16:22:56.000000000 +0200
+++ gnome-control-center-2.30.0/capplets/about-me/gnome-about-me.c	2010-03-29 17:03:35.768549194 +0200
@@ -29,6 +29,7 @@
 #include <unistd.h>
 #include <libebook/e-book.h>
 #include <dbus/dbus-glib-bindings.h>
+#include <unique/unique.h>
 
 #define GNOME_DESKTOP_USE_UNSTABLE_API
 #include <libgnomeui/gnome-desktop-thumbnail.h>
@@ -835,7 +836,7 @@
 				    me->disable_fingerprint_button);
 }
 
-static gint
+static GtkWindow*
 about_me_setup_dialog (void)
 {
 	GtkWidget    *widget;
@@ -856,7 +857,7 @@
 
 	if (dialog == NULL) {
 		about_me_destroy ();
-		return -1;
+		return NULL;
 	}
 
 	me->dialog = dialog;
@@ -891,7 +892,7 @@
 						"Evolution Data Server can't handle the protocol"));
 			g_clear_error (&error);
 			about_me_destroy ();
-			return -1;
+			return NULL;
 		}
 
 		g_clear_error (&error);
@@ -979,27 +980,80 @@
 
 	gtk_widget_show_all (main_dialog);
 
-	return 0;
+	return GTK_WINDOW (main_dialog);
+}
+
+static UniqueResponse
+message_received_cb (UniqueApp         *app,
+		     UniqueCommand      command,
+		     UniqueMessageData *message,
+		     guint              time,
+		     gpointer           user_data)
+{
+	UniqueResponse  res;
+	GtkWindow   *main_dialog;
+	
+	main_dialog = GTK_WINDOW (user_data);
+
+	switch (command) {
+	case UNIQUE_ACTIVATE:
+		/* move the main window to the screen that sent us the command */
+		/* FIXME: This does not appear to be working */
+		gtk_window_set_screen (main_dialog,
+				       unique_message_data_get_screen (message));
+		
+		gtk_window_present_with_time (main_dialog, time);
+
+		res = UNIQUE_RESPONSE_OK;
+		break;
+	default:
+		res = UNIQUE_RESPONSE_PASSTHROUGH;
+		break;
+	}
+
+	return res;
 }
 
 int
 main (int argc, char **argv)
 {
-	int rc = 0;
-
+	GtkWindow *main_dialog;
+	UniqueApp *unique_app;
+	
 	capplet_init (NULL, &argc, &argv);
 
 	if (!g_thread_supported ())
 		g_thread_init (NULL);
+	
+	unique_app = unique_app_new ("org.gnome.gnome-about-me", NULL);
+	if (unique_app_is_running (unique_app)) {
+		int retval = 0;
+
+		UniqueResponse response;
+		response = unique_app_send_message (unique_app,
+						    UNIQUE_ACTIVATE,
+						    NULL);
+		retval = (response != UNIQUE_RESPONSE_OK);
+
+		g_object_unref (unique_app);
+		return retval;
+	}
 
 	dbus_g_object_register_marshaller (fprintd_marshal_VOID__STRING_BOOLEAN,
-					   G_TYPE_NONE, G_TYPE_STRING, G_TYPE_BOOLEAN, G_TYPE_INVALID);
+					   G_TYPE_NONE, G_TYPE_STRING,
+					   G_TYPE_BOOLEAN, G_TYPE_INVALID);
 
-	rc = about_me_setup_dialog ();
+	main_dialog = about_me_setup_dialog ();
 
-	if (rc != -1) {
+	if (main_dialog != NULL) {
+		unique_app_watch_window (unique_app, main_dialog);
+		g_signal_connect (unique_app, "message-received",
+				  G_CALLBACK (message_received_cb), main_dialog);
 		gtk_main ();
+		return 0;
+	} else {
+		g_critical ("Failed to build main dialog");
+		return 1;
 	}
 
-	return rc;
 }
